# NLST command should interpret paths starting with './' as relative
# with respect to current working directory.
#
# running ftp NLST command we obtain different results
# for pattern 'test/TEST????.dat' and './test/TEST????.dat'
#
#    ftp> ls test/TEST????.dat
#    200 PORT command successful
#    150 Opening ASCII mode data connection for file list
#    test/TEST0001.dat
#    test/TEST0006.dat
#    test/TEST0008.dat
#    test/TEST0009.dat
#    test/TEST0007.dat
#    test/TEST0000.dat
#    test/TEST0003.dat
#    test/TEST0004.dat
#    test/TEST0005.dat
#    test/TEST0002.dat
#    226 Transfer complete
#    remote: test/TEST????.dat
#    190 bytes received in 0.0033 seconds (55.61 Kbytes/s)
#    ftp> ls ./test/TEST????.dat
#    200 PORT command successful
#    150 Opening ASCII mode data connection for file list
#    226 Transfer complete
#
# same change in mod_ls.c has been submitted in pull request here:
# 	https://github.com/proftpd/proftpd/pull/1283
#
# patch also combines a change to is_dotdir(), so it makes it identical
# with is_dotdir() as found in master branch
#

--- a/modules/mod_ls.c
+++ b/modules/mod_ls.c
@@ -3182,10 +3182,6 @@ MODRET ls_nlst(cmd_rec *cmd) {
       p = *path;
       path++;
 
-      if (*p == '.' && (!opt_A || is_dotdir(p))) {
-        continue;
-      }
-
       pr_fs_clear_cache2(p);
       if (pr_fsio_stat(p, &st) == 0) {
         /* If it's a directory... */
diff --git a/src/dirtree.c b/src/dirtree.c
index 4f6b449da..c12347235 100644
--- a/src/dirtree.c
+++ b/src/dirtree.c
@@ -98,10 +98,10 @@ static int allow_dyn_config(const char *path) {
 
 /* Return true if dir is ".", "./", "../", or "..". */
 int is_dotdir(const char *dir) {
-  if (strncmp(dir, ".", 2) == 0 ||
-      strncmp(dir, "./", 2) == 0 ||
-      strncmp(dir, "..", 3) == 0 ||
-      strncmp(dir, "../", 3) == 0) {
+  if (strcmp(dir, ".") == 0 ||
+      strcmp(dir, "./") == 0 ||
+      strcmp(dir, "..") == 0 ||
+      strcmp(dir, "../") == 0) {
     return TRUE;
   }
 
